### DISCUSSION TITLE
Swift implementation of NARS

### DISCUSSION THREAD

**Discussion by maxeeem** [AGI-course/agi-course.github.io#4] [Ideas]
I am working on a [Swift implementation of a NARS system](https://github.com/maxeeem/NARS-Swift).

Primarily I am doing this for pedagogical reasons as I am learning the theory. 

**Some goals of this project:**
- small and simple codebase
- cross-platform
- embedding Narsese as a DSL in Swift
- using miniKanren for pattern matching
- providing a Swift Playground for easy experimentation

At this point the lower layers are functioning pretty well and I am currently working on temporal reasoning. There are many TODOs in the code for future improvements as I am currently focusing more on functionality than performance. If you have any comments or suggestions I would love to connect and discuss!

**Comment by patham9** [comment]
This is an ambitious project, but due to your background with NARS also very fitting!
Do you have certain domains in mind where you plan to test the capabilities of your system?
Also, I wonder if there are aspects of other AI systems you find potentially valuable (such as other AGI systems and ML models) which could enhance this implementation when integrated or interfaced with.

**Comment by maxeeem** [comment]
After I add Nal-7&8 and the system performs as expected I want to apply it to natural language. Along the lines of https://cis.temple.edu/~pwang/Publication/NLP.pdf I may even start smaller and begin with the alphabet like you teach children. In my view Nars is like a child and so similar teaching methods should apply. That's the theory at least :)

As for other AGI systems - absolutely! Code from OpenNARS and ONA have been invaluable in learning the system. I took inspiration from your rule definition for example. I also ported test cases from OpenNARS to make sure that my system is comparable in capabilities. 

When it comes to ML I think it will be part of a complete system, what Pei calls NARS+. The reasoner will interface with various channels like your examples with Transbot and YOLO but those are not going to be part of the core system. I am however using miniKanren to do pattern matching when applying rules and this seems to work really well. 

In the future I hope our systems can all interface with each other and we should be able to test them on the same tasks. As long as they all "speak" Narsese this should be possible.

**Reply by patham9** [reply]
Sounds all plausible. In addition I encourage everyone to also look at other systems than NARS despite it being the only system we chose to capture within the only 8 AGI lectures.
Some which have implementations which are still being worked on:
AERA - https://openaera.org/
OpenCog - https://opencog.org/
ACT-R - http://act-r.psy.cmu.edu/software/
